pallets/connectors/src/routers.rs

```diff
@@ -22,9 +22,34 @@ pub struct XcmDomain<CurrencyId> {
 	/// It should contain the pallet index + the `transact` call index, to which
 	/// we will append the eth_tx param. You can obtain this value by building
 	/// an ethereum_xcm::transact call with Polkadot JS on the target chain.
-	pub ethereum_xcm_transact_call_index: Vec<u8>,
+	pub ethereum_xcm_transact_call_index:
+		BoundedVec<u8, ConstU32<{ xcm_primitives::MAX_ETHEREUM_XCM_INPUT_SIZE }>>,
 	/// The ConnectorsXcmRouter contract address on a given domain
 	pub contract_address: H160,
 	/// The currency in which execution fees will be paid on
 	pub fee_currency: CurrencyId,
 }
+
+// NOTE: Remove this custom implementation once the following underlying data implements MaxEncodedLen:
+/// * Polkadot Repo: xcm::VersionedMultiLocation
+/// * PureStake Repo: pallet_xcm_transactor::Config<Self = T>::CurrencyId
+impl<CurrencyId> MaxEncodedLen for XcmDomain<CurrencyId>
+where
+	XcmDomain<CurrencyId>: Encode,
+{
+	fn max_encoded_len() -> usize {
+		// The domain's `VersionedMultiLocation` (custom bound)
+		xcm::v1::MultiLocation::max_encoded_len()
+			// From the two enum variants of `VersionedMultiLocation
+			.saturating_add(2)
+			// The ethereum xcm call index (default bound)
+			.saturating_add(BoundedVec::<
+				u8,
+				ConstU32<{ xcm_primitives::MAX_ETHEREUM_XCM_INPUT_SIZE }>,
+			>::max_encoded_len())
+			// The contract address (default bound)
+			.saturating_add(H160::max_encoded_len())
+			// The fee currency (custom bound)
+			.saturating_add(cfg_types::tokens::CurrencyId::max_encoded_len())
```

Comment:
TBH, I have no clue how the `MaxEncodedLen` is derived. For an enum I would guess, assuming an enum has less then 256 variants, it would be 1 bytes + size of largest enum variant? 

As the differences between `v0` and `v1` are basically 
* `v1::MultiLocation` being a struct holding a `u8` and essentially `v0::MultiLocation`
* `v1::Junction` removing the named variant (`Parent`) and all other items being equal

I would say the implementation is correct. WDYT @NunoAlexandre?

The only thing I don't understand is why we need to add 2, wouldn't it be 1 as in 1 byte for indicating the variant number in the codec?

Reply:
Maybe I don't understand `MaxEncodedLength` but I thought its point was to capture the max length that a value of a specific type could take; for example for the  If that is correct, than a `VersionedMultiLocation` will not be at most 2 bytes, it can in fact be arbitrarily long and that's probably why `MaxEncodedLength` is not implemented for that type.

[Help me understand](https://www.youtube.com/watch?v=tGu0_rFwfBc) 😆 🙏

Reply:
> The only thing I don't understand is why we need to add 2, wouldn't it be 1 as in 1 byte for indicating the variant number in the codec?

Yes, you are right. I mixed that up with something else. Validated this with a small example:

```
#[derive(codec::Encode, codec::MaxEncodedLen)]
pub enum TestEnum<T: MaxEncodedLen> {
	A(T),
	B(T),
        C(T),
        D(T),
        E(T),
}

#[test]
assert_eq!(TestEnum::<u32>::max_encoded_len(), u32::max_encoded_len() + 1);
```

> Maybe I don't understand `MaxEncodedLength` but I thought its point was to capture the max length that a value of a specific type could take; If that is correct, then a `VersionedMultiLocation` will not be at most 2 bytes, it can in fact be arbitrarily long and that's probably why `MaxEncodedLength` is not implemented for that type.


`MaxEncodingLen` aims to provide an estimate to the maximum encoded size of an arbitrary storage. So even though a `u128` could have length 1, its [max encoded length would be 17](https://github.com/paritytech/parity-scale-codec/blob/master/src/max_encoded_len.rs#L75). Some more context can be found [in the initial PR which introduced it](https://github.com/paritytech/substrate/issues/8719). 

Unfortunately, [`VersionedMultiLocation`](https://github.com/paritytech/polkadot/blob/53253de77f1649ecbb650df2ec3e29d0d1110a2d/xcm/src/lib.rs#L73) does not impl `MaxEncodedLen`.

```
pub enum VersionedMultiLocation {
	V0(v0::MultiLocation),
	V1(v1::MultiLocation),
}
```

Luckily, `v1::MultiLocation` does! Since [v1 ](https://github.com/paritytech/polkadot/blob/53253de77f1649ecbb650df2ec3e29d0d1110a2d/xcm/src/v1/multilocation.rs#L51) is [v0](https://github.com/paritytech/polkadot/blob/53253de77f1649ecbb650df2ec3e29d0d1110a2d/xcm/src/v0/multi_location.rs#L46) with an extra field as pointed out above, v1 should be fine for the `MaxEncodedLen` plus the encoding size of an enum which is one.

 Therefore, it should be correct to use `v1::Multilocation::max_encoding_len().saturating_add(1)` for `VersionedMultiLocation::max_encoded_len()`.

Reply:
Oh I see, now I get it 🙏 thanks! Looks good then :)